onnxruntime/core/optimizer/embed_layer_norm_fusion.cc

```diff
@@ -419,66 +562,65 @@ Status EmbedLayerNormFusion::ApplyImpl(Graph& graph, bool& modified, int graph_l
     auto wg_shape = word_embedding->Shape();
     if (wg_shape == nullptr || wg_shape->dim_size() != 2 ||
         !utils::HasDimValue(wg_shape->dim()[1]) ||
-        wg_shape->dim()[1].dim_value() != sg_shape->dim()[1].dim_value()) {
-      continue;
-    }
-
-    // Traceback the Add node to find (Shape --> Expand -->) Gather --> Add.
-    // Constant folding removes Shape and Expand nodes when input does not have symbolic shape. In that
-    // case just look for Gather --> Add.
-    std::vector<graph_utils::EdgeEndToMatch> position_embedding_path{
-        {0, 1, "Gather", {1, 11}, kOnnxDomain}};
-    if (!graph_utils::FindPath(add_node, true, position_embedding_path, edges, logger)) {
-      continue;
-    }
-    Node& position_gather_node = *graph.GetNode(edges[0]->GetNode().Index());
-    if (position_gather_node.GetOutputEdgesCount() != 1) {
-      continue;
-    }
-    // The first input of position_gather_node must be 2d.
-    NodeArg* position_embedding = position_gather_node.MutableInputDefs()[0];
-    auto pg_shape = position_embedding->Shape();
-    if (pg_shape == nullptr || pg_shape->dim_size() != 2 ||
-        !utils::HasDimValue(pg_shape->dim()[1]) ||
-        pg_shape->dim()[1].dim_value() != sg_shape->dim()[1].dim_value()) {
+        wg_shape->dim()[1].dim_value() != hidden_size) {
       continue;
     }
 
     NodeArg* input_ids = word_gather_node.MutableInputDefs()[1];
+    NodeArg* position_embedding = nullptr;
+    std::vector<NodeIndex> nodes_to_remove;
+
+    auto add_input_name = add_node.MutableInputDefs()[1]->Name();
+    if (graph_utils::IsConstantInitializer(graph, add_input_name)) {
+      // Check that input has static shape
+      auto input_shape = input_ids->Shape();
+      if (input_shape->dim_size() != 2 ||
+          !utils::HasDimValue(input_shape->dim()[0]) ||
+          !utils::HasDimValue(input_shape->dim()[1])) {
+        continue;
+      }
+
+      // check tensor shape [batch_size, sequence_length, hidden_size]
+      int64_t batch_size = input_shape->dim()[0].dim_value();
+      int64_t sequence_length = input_shape->dim()[1].dim_value();
+      int64_t input_size = batch_size * sequence_length;
```

Comment:
so you only fuse when input has static shape here?

Reply:
This case only happens for static shape due to constant folding.